signer/src/keys.rs

```diff
@@ -0,0 +1,369 @@
+        value.0
+    }
+}
+
+impl From<&PublicKey> for secp256k1::XOnlyPublicKey {
+    fn from(value: &PublicKey) -> Self {
+        value.0.x_only_public_key().0
+    }
+}
+
+impl From<&PublicKey> for p256k1::point::Point {
+    /// This implementation takes the full 65 byte serialization of the
+    /// public key and breaks it into it's x-coordinate and y-coordinate
+    /// parts, and then maps those coordinates into a Point.
+    ///
+    /// # Notes
+    ///
+    /// An uncompressed serialization of the secp256k1::PublicKey type is
+    /// 65 bytes. The first byte denotes that the following slice is an
+    /// uncompressed public key on the secp256k1 curve, the next 32 bytes
+    /// are for the x-coordinate, and the remaining 32-bytes are for the
+    /// y-coordinate.
+    fn from(value: &PublicKey) -> Self {
+        // We start by serializing the full key into it's x- and
+        // y-coordinates.
+        let full_key: [u8; 65] = value.0.serialize_uncompressed();
+
+        // Let's copy over the various slices. The copy cannot panic
+        // because we know that the lengths of each of the slices match.
+        let mut x_part = [0; 32];
+        let mut y_part = [0; 32];
+        x_part.copy_from_slice(&full_key[1..33]);
+        y_part.copy_from_slice(&full_key[33..]);
+
+        // Okay, now for conversion to the p256k1 types. Under the hood
+        // here `p256k1::field::Element::from` tries to reduce the input to
+        // the order of the secp256k1 curve[1][2], but we do not need worry
+        // since we have a valid point on the curve.
+        //
+        // [^1]: https://github.com/Trust-Machines/p256k1/blob/3ecb941c1af13741d52335ef911693b6d6fda94b/p256k1/src/field.rs#L268-L279
+        // [^2]: https://github.com/bitcoin-core/secp256k1/blob/v0.3.0/src/field.h#L78-L79
+        let x_element = p256k1::field::Element::from(x_part);
+        let y_element = p256k1::field::Element::from(y_part);
+        // You would think that you couldn't always convert two elements
+        // into a Point, but `p256k1::point::Point::from` uses
+        // `secp256k1_gej_set_ge` under the hood, which I believe does any
+        // reduction. But still, we have a valid public key, so no
+        // reductions should be necessary.
+        p256k1::point::Point::from((x_element, y_element))
+    }
+}
+
+/// This should only error when the `p256k1::point::Point` is the identity
+/// point (also called the at infinity).
+impl TryFrom<&p256k1::point::Point> for PublicKey {
+    type Error = Error;
+    fn try_from(value: &p256k1::point::Point) -> Result<Self, Self::Error> {
+        let x_data = value.x().to_bytes();
+
+        let pk = secp256k1::XOnlyPublicKey::from_slice(&x_data).map_err(Error::InvalidPublicKey)?;
+        let parity = if value.has_even_y() {
+            Parity::Even
+        } else {
+            Parity::Odd
+        };
+
+        let public_key = secp256k1::PublicKey::from_x_only_public_key(pk, parity);
```

Comment:
Remind me, why is the parity necessary?

Reply:
It’s necessary for the 32 byte representation of an XOnlyPublicKey, because there are two points on the curve with the same x value.